src/shared/serialize.c

```diff
@@ -255,6 +279,42 @@ int deserialize_environment(const char *value, char ***list) {
         return 0;
 }
 
+int deserialize_pidref(FDSet *fds, const char *value, PidRef *ret) {
+        const char *e;
+        int r;
+
+        assert(value);
+        assert(ret);
+
+        e = startswith(value, "@");
```

Comment:
I'd prefer to have separate keys, and always send both, and have the pidfd k/v win over the other one, rather than encoding magic specifiers, which is more opaque

Reply:
no, i decided against that, since that means we cannot do upgrades robustly, because we'd have to serialize in an order (i.e. we'd strictly have to serialize the pidfd first, and the pid second, so that we never act on the pid before we use the pidfd).

moreover, it's messy then because this would basically double the deserialization logic, since we always have to add two keys to the big streq() if series.

it's a *ton* simpler this way, shorter than before in fact. And we do not serialize redundant data, but still retain compat when upgrading from old version. It's a *ton* better this way I am sure.

my patch initially did what you said, but then i scrapped it, because it made everything messier so much.